### ISSUE TITLE
Contract dependencies binding

### ISSUE THREAD

**Issue by lgalabru** [blockstack/blockstack-core#1111]
While working on #1105, I noticed an issue / limitation.
When de-sugaring principals in the context of `fetch-contract-entry` and `contract-call!` is straight forward, 
```
(contract-map-get contract-name kv-store ((key 1))) -> fetch-contract-entry SP00...GH.contract-name kv-store ((key "1")))
```

de-sugaring a principal being used for instance, in the context of a tuple, is a different story:
```
(map-get kv-store ((key contract-name))) -> we need to know that "key" is a Principal for de-sugaring "contract-name"
```

The 2 solutions I can think of are:
1) the de-sugaring is no longer a mutating pass executed post-parsing. It happens on the fly during the type-checking pass and at runtime. In other words, I think we cancel the `early binding of names to contract` feature.
2) we explicitly list dependencies (as described in https://github.com/blockstack/blockstack-core/issues/1078 - or differently), and the de-sugaring pass is naively expending the contract names occurrences:
```
(define-dependency rocket-exchange ((name "rocket-exchange") (principal 'SP0...0001) (version 1)))
(define-dependency rocket-token  ((name "rocket-token")  (principal 'SP0...0001) (version 1)))
```

Am I missing a dead angle? Any thoughts?

**Comment by jcnelson** [comment]
> the de-sugaring is no longer a mutating pass executed post-parsing. It happens on the fly during the type-checking pass and at runtime. In other words, I think we cancel the early binding of names to contract feature.

I think there's still value in doing a pass at publish time as to whether or not each `contract-call!`'s contract name resolves.  However, I don't think I understand the concern?  There isn't a type that encodes a fully-qualified contract name (AFAIK), so I don't think the `kv-store` example you listed was ever expected to be an issue.

**Comment by kantai** [comment]
I think this could be solved by some specification of the syntax for contract name usage.

Currently, `contract-call!` and `fetch-contract-entry` treat the contract-name parameter as a "name", e.g.:

```
(contract-call! my-other-contract foo)
```

Whereas a contract principal Value is a _literal_ which then becomes a Clarity Value object:

```
(tuple (key 'CTmy-other-contract))
```

That exact syntax probably isn't viable with #1078, and in any case would lead to a lot of special case handling in the sugar-expander.

However, something like the following:

```
(contract-call! .other-contract foo)
(tuple (key .other-contract))
```

Is readily parseable, and presents a viable strategy for a sugar-expander. Namely:

0. Add a `LiteralValue` variant to `SymbolicRepresentation` -- this is identical to AtomValue, but will be used to signify that a value was an actual literal. It may only be constructed in the parser passes.
1. `parser` produces a new enum type PreSymbolicRepresentation AST with an element for unexpanded contract principals.
2. `sugar-expander` produces the normal SymbolicRepresentation AST, with all unexpanded contract principals expanded into `LiteralValues`s
3. TypeChecker enforces that the second argument to `contract-call!` and `fetch-contract-entry` is (1) a LiteralValue of a Contract Principal, (2) exactly known at compile time, (3) and then performs the type analysis using that contract principal.
4. The VM/interpreter also must perform a check that the symbolic representation is a LiteralValue.

**Comment by lgalabru** [comment]
Addressed with #1105
